smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/HttpProtocolTestGenerator.java

```diff
@@ -377,13 +456,33 @@ private void writeHttpRequestAssertions(HttpRequestTestCase testCase) {
 
         writeHttpHeaderAssertions(testCase);
         writeHttpQueryAssertions(testCase);
-        writeHttpBodyAssertions(testCase);
+        testCase.getBody().ifPresent(body -> {
+            writeHttpBodyAssertions(body, testCase.getBodyMediaType().orElse("UNKNOWN"), true);
+        });
     }
 
     private void writeHttpResponseAssertions(HttpResponseTestCase testCase) {
         writer.write("expect(r.statusCode).toBe($L);", testCase.getCode());
         writeHttpHeaderAssertions(testCase);
-        writeHttpBodyAssertions(testCase);
+        testCase.getBody().ifPresent(body -> {
+            writeHttpBodyAssertions(body, testCase.getBodyMediaType().orElse("UNKNOWN"), false);
+        });
+    }
+
+    private void writeHttpResponseAssertions(HttpMalformedResponseDefinition responseDefinition) {
+        writer.write("expect(r.statusCode).toBe($L);", responseDefinition.getCode());
+        responseDefinition.getHeaders().forEach((header, value) -> {
+            header = header.toLowerCase();
+            writer.write("expect(r.headers[$S]).toBeDefined();", header);
+            writer.write("expect(r.headers[$S]).toBe($S);", header, value);
+        });
+        writer.write("");
+        responseDefinition.getBody().ifPresent(body -> {
+            writeHttpBodyAssertions(body, responseDefinition.getBodyMediaType().orElse("UNKNOWN"), false);
+        });
+        responseDefinition.getBodyMessageRegex().ifPresent(regex -> {
+            writeHttpBodyMessageAssertion(regex, responseDefinition.getBodyMediaType().orElse("UNKNOWN"));
+        });
```

Comment:
Is there ever any need to generate both of these?

Reply:
Upstream it is asserted you cannot have both of these. So only one will ever be written.

Reply:
And now in the trait these two are members of a union.